modules/040-node-manager/hooks/yc_delete_preemptible_instances.go

```diff
@@ -0,0 +1,365 @@
+		Name:  obj.GetName(),
+		Nodes: nodeCount,
+		Ready: readyNodeCount,
+	}, nil
+}
+
+func applyNodeFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	labels := obj.GetLabels()
+
+	ng, ok := labels["node.deckhouse.io/group"]
+	if !ok {
+		return nil, nil
+	}
+
+	return &Node{
+		Name:              obj.GetName(),
+		NodeGroup:         ng,
+		CreationTimestamp: obj.GetCreationTimestamp(),
+	}, nil
+}
+
+func isPreemptibleFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	preemptible, ok, err := unstructured.NestedBool(obj.UnstructuredContent(), "spec", "schedulingPolicy", "preemptible")
+	if err != nil {
+		return nil, fmt.Errorf("can't access field \"spec.schedulingPolicy.preemptible\" of YandexMachineClass %q: %s", obj.GetName(), err)
+	}
+
+	if ok && preemptible {
+		return &YandexMachineClass{
+			Name: obj.GetName(),
+		}, nil
+	}
+
+	return nil, nil
+}
+
+var _ = sdk.RegisterFunc(&go_hook.HookConfig{
+	AllowFailure: true,
+	Queue:        "/modules/cloud-provider-yandex/preemtibly-delete-preemtible-instances",
+	Schedule: []go_hook.ScheduleConfig{
+		{
+			Name: "every-15",
+			// string formatting is ugly, but serves a purpose of referencing an important constant
+			Crontab: fmt.Sprintf("0/%.0f * * * *", hookExecutionSchedule.Minutes()),
+		},
+	},
+	Kubernetes: []go_hook.KubernetesConfig{
+		{
+			Name:                "mcs",
+			ExecuteHookOnEvents: go_hook.Bool(false),
+			ApiVersion:          "machine.sapcloud.io/v1alpha1",
+			Kind:                "YandexMachineClass",
+			NamespaceSelector: &types.NamespaceSelector{
+				NameSelector: &types.NameSelector{
+					MatchNames: []string{"d8-cloud-instance-manager"},
+				},
+			},
+			FilterFunc: isPreemptibleFilter,
+		},
+		{
+			Name:                "machines",
+			ExecuteHookOnEvents: go_hook.Bool(false),
+			ApiVersion:          "machine.sapcloud.io/v1alpha1",
+			Kind:                "Machine",
+			NamespaceSelector: &types.NamespaceSelector{
+				NameSelector: &types.NameSelector{
+					MatchNames: []string{"d8-cloud-instance-manager"},
+				},
+			},
+			FilterFunc: applyMachineFilter,
+		},
+		{
+			Name:                "nodes",
+			ExecuteHookOnEvents: go_hook.Bool(false),
+			ApiVersion:          "v1",
+			Kind:                "Node",
+			FilterFunc:          applyNodeFilter,
+		},
+		{
+			Name:                "nodegroupstatuses",
+			ExecuteHookOnEvents: go_hook.Bool(false),
+			ApiVersion:          "deckhouse.io/v1",
+			Kind:                "NodeGroup",
+			FilterFunc:          applyNodeGroupFilter,
+		},
+	},
+}, deleteMachines)
+
+func deleteMachines(input *go_hook.HookInput) error {
+	var (
+		timeNow                        = time.Now().UTC()
+		machines                       []*Machine
+		preemptibleMachineClassesSet   = set.Set{}
+		nodeNameToNodeMap              = make(map[string]*Node)
+		nodeGroupNameToNodeGroupStatus = make(map[string]*NodeGroupStatus)
+	)
+
+	for _, mcRaw := range input.Snapshots["mcs"] {
```

Comment:
Consider 
https://pkg.go.dev/github.com/deckhouse/deckhouse@v1.43.0/go_lib/set#NewFromSnapshot

Reply:
I dislike this interface. It does not provide proper empty (nil or zero) string value filtering.

Reply:
Then let's add nils filtering to the NewFromSnapshot function.

Reply:
Added!